lib/operations/common/GitlabRepoRef.ts

```diff
@@ -0,0 +1,157 @@
+import * as url from "url";
+import {
+    ActionResult,
+    successOn,
+} from "../../action/ActionResult";
+import { Configurable } from "../../project/git/Configurable";
+import { DefaultHttpClientFactory } from "../../spi/http/axiosHttpClient";
+import { HttpMethod } from "../../spi/http/httpClient";
+import { logger } from "../../util/logger";
+import { AbstractRemoteRepoRef } from "./AbstractRemoteRepoRef";
+import { GitlabPrivateTokenCredentials } from "./GitlabPrivateTokenCredentials";
+import { GitShaRegExp } from "./params/validationPatterns";
+import { ProjectOperationCredentials } from "./ProjectOperationCredentials";
+import { ProviderType } from "./RepoId";
+
+export const GitlabDotComApiBase = "https://gitlab.com/api/v4";
+export const GitlabDotComRemoteUrl = "https://gitlab.com/";
+
+/**
+ * Repository reference implementation for Gitlab
+ */
+export class GitlabRepoRef extends AbstractRemoteRepoRef {
+
+    public static from(params: {
+        owner: string,
+        repo: string,
+        sha?: string,
+        rawApiBase?: string,
+        path?: string,
+        gitlabRemoteUrl?: string,
+        branch?: string,
+    }): GitlabRepoRef {
+        if (params.sha && !params.sha.match(GitShaRegExp.pattern)) {
+            throw new Error("You provided an invalid SHA: " + params.sha);
+        }
+        const result = new GitlabRepoRef(params.owner, params.repo, params.sha, params.rawApiBase, params.gitlabRemoteUrl, params.path);
+        result.branch = params.branch;
+        return result;
+    }
+
+    private constructor(owner: string,
+                        repo: string,
+                        sha: string,
+                        public apiBase = GitlabDotComApiBase,
+                        gitlabRemoteUrl: string = GitlabDotComRemoteUrl,
+                        path?: string) {
+        super(apiBase === GitlabDotComApiBase ? ProviderType.gitlab_com : ProviderType.gitlab_enterprise,
+            gitlabRemoteUrl,
+            apiBase,
+            owner,
+            repo,
+            sha,
+            path);
+    }
+
+    public async createRemote(creds: ProjectOperationCredentials, description: string, visibility): Promise<ActionResult<this>> {
+
+        const gitlabUrl = this.concatUrl(this.apiBase, `projects`);
+        const httpClient = DefaultHttpClientFactory.create();
```

Comment:
That's not how the HttpClientFactory is supposed to be accessed as it ties it to one particular implementation. Probably better to get it from the config like `automationClientInstance().configuration.http.client.factory`.